---
title: "Grover Gene Expression Data: Dimension Reduction"
author: "Michelle Li and Gianna Wu with Professor de Pillis and Dr. Park"
date: "7/20/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(destiny)
library(Biobase)
library(tsne)
library(rgl)
options(rgl.useNULL = TRUE)
r3dDefaults$windowRect <- c(0,0,1000,1000)
```

```{r load-grover, include=FALSE}
## load .Rdata file into environment so you have all calculations done (located in our Drive)
## if you have the .RData file to load, use eval=FALSE for pca, diffmap, tsne
load("./grover-rmd-data.RData")
```

----

Analyzing Grover gene expression data using the following dimension reduction techniques in R: PCA (handwritten version), diffusion mapping (destiny package), and t-SNE (tsne package). The original Grover data comes in as 46175 genes x 135 cells.  

Grover looks at how changes in properties of stem cells relates to mammalian aging and physiological changes. They do single-cell analysis of hematopoeitic stem cells (HSCs) in young mice (2-3 months old) and old mice (20-25 months old).   

```{r prepare-grover, eval=FALSE, include=FALSE}
## read in data, convert to matrix
rawdata <- read.csv("./data/grover_expression.txt", row.names=1, sep="")
data <- as.matrix(rawdata)

## grouping
cellnames <- as.character(colnames(data))
celltypes <- c("A10_young","A11_young", "A3_young", "A4_young", "A5_young", "A7_young", "A9_young", "B11_young", "B12_young", "B1_young", "B2_young", "B3_young", "B5_young", "B7_young", "B8_young", "C11_young", "C12_young", "C1_young", "C3_young", "C4_young", "C5_young", "C6_young", "C7_young", "C8_young", "D1_young", "D3_young", "D6_young", "D7_young", "D8_young", "D9_young", "E11_young", "E1_young", "E2_young", "E3_young", "E6_young", "E7_young", "F10_young", "F1_young", "F2_young", "F3_young", "F5_young", "F6_young", "F8_young", "F9_young", "G11_young", "G12_young", "G1_young", "G2_young", "G3_young", "G4_young", "G5_young", "G6_young", "G7_young", "G8_young", "H11_young", "H12_young", "H3_young", "H4_young", "H5_young", "H6_young", "H8_young", "A10_old", "A11_old", "A12_old", "A1_old", "A2_old", "A4_old", "A5_old", "A6_old", "A7_old", "A8_old", "B10_old", "B11_old", "B2_old", "B3_old", "B4_old", "B5_old", "B6_old", "B7_old", "B9_old", "C10_old", "C11_old", "C12_old", "C1_old", "C3_old", "C4_old", "C5_old", "C6_old", "C7_old", "C8_old", "C9_old", "D11_old", "D12_old", "D1_old", "D4_old", "D6_old", "D7_old", "D8_old", "D9_old", "E10_old", "E11_old", "E12_old", "E1_old", "E3_old", "E5_old", "E7_old", "E8_old", "E9_old", "F10_old", "F12_old", "F1_old", "F3_old", "F4_old", "F5_old", "F7_old", "F8_old", "F9_old", "G10_old", "G12_old", "G1_old", "G2_old", "G3_old", "G5_old", "G6_old", "G9_old", "H11_old", "H12_old", "H1_old", "H2_old", "H4_old", "H5_old", "H6_old", "H7_old", "H8_old", "H9_old")

## create factor to cluster based on age
age <- celltypes
age[grepl("young",age)] <- "young"
age[grepl("old",age)] <- "old"
age <- factor(age)

# color
colors <- c("limegreen", "deeppink1")
```

```{r heatmap-grover, eval=FALSE, include=FALSE}
hmap <- heatmap(data, Rowv=NA, Colv=NA)
```


## PCA (Handwritten)

This is the handwritten version of PCA in R, I used SVD for everything (not eig).

```{r pca-grover, eval=FALSE, include=FALSE}
## PCA DIMENSIONS:
## cells = rows, genes = columns 
## for Grover data: use TRANSPOSE of data matrix (want cells x genes)
data <- t(data)
M1centered <- scale(data, center=TRUE, scale=FALSE) # subtracts column means from each column

## svd
svd <- svd(M1centered, nu=3, nv=3) # only first 3 left/right singular vectors for efficiency
U <- svd$u # dim: nrowx3
V <- svd$v # dim: ncolx3
## Note: U = M1centered %*% V %*% S^-1

## calculate scores
# weights <- t(M1centered) %*% U # dim: ncolx3
scores <- M1centered %*% V # dim: nrowx3
```


#### Plots of scores
Scores were calculated by: scores <- data %*% V. V are the right singular vectors from SVD whose columns contain the eigenvectors of the covariance matrix. Scores give the coordinates of the individuals on the principal components (coefficients of the linear combination of original variables for each principal component). 

2D plot of scores  
```{r pca-plot2dscores}
plot(x=scores[,1], y=scores[,2], 
     xlab="PC1 scores", ylab="PC2 scores", 
     col = colors[age],
     pch=20, 
     main = "PCA Scores, Grover")
legend("topright", inset = c(0,0),
       legend = levels(age),
       pch = 20,
       col = colors,
       ncol=1, cex=0.6, pt.cex=1)
```

3D plot of scores  
```{r pca-plot3dscores}
## plot scores 3d
plot3d(x=scores[,1], y=scores[,2], z=scores[,3], 
       xlab="PC1 scores", ylab="PC2 scores", zlab="PC3 scores", 
       col = colors[age],
       pch=20, 
       main = "PCA Scores, Grover")
legend3d("bottomright", inset=c(0.2,0.2),
         legend = levels(age),
         pch = 20,
         col = colors,
         ncol=1, cex=1.5)
rglwidget()
```

#### Plots of principal components
2D plot of PCs (eigenvectors of covariance matrix)  
```{r pca-plot2dpcs}
## plot PCs 2d (eigenvectors of covariance matrix)
plot(x=V[,1], y=V[,2], 
     xlab="PC1", ylab="PC2", 
     col = colors[age],
     pch=20, 
     main = "Principal Components, Grover")
legend("topright", inset = c(0,0),
       legend = levels(age),
       pch = 20,
       col = colors,
       ncol=1, cex=0.6, pt.cex=1)
```

2D plot of PCs (zoomed in)  
```{r pca-plot2dpcs-zoom}
## plot PCs 2d (eigenvectors of covariance matrix)
plot(x=V[,1], y=V[,2], 
     xlab="PC1", ylab="PC2", 
     xlim = c(-1.5e-3, 1.5e-3), ylim = c(-.01, 5e-3),
     col = colors[age],
     pch=20, 
     main = "Principal Components, Grover")
legend("bottomleft", inset = c(0,0),
       legend = levels(age),
       pch = 20,
       col = colors,
       ncol=1, cex=0.6, pt.cex=1)
```

3D plot of PCs (eigenvectors of covariance matrix)  
```{r pca-plot3d-pcs}
## plot PCs 3d
plot3d(x=V[,1], y=V[,2], z=V[,3], 
       xlab="PC1", ylab="PC2", zlab="PC3", 
       col = colors[age],
       pch=20, 
       main = "Principal Components, Grover")
legend3d("bottomright", inset=c(0.2,0.2),
         legend = levels(age),
         pch = 20,
         col = colors,
         ncol=1, cex=1.5)
rglwidget()
```

----

## Diffusion Mapping (destiny)

Diffusion mapping using destiny package.  
```{r diffmap-grover, eval=FALSE, include=FALSE}
## convert to expression set
## EXPRESSION SET DIMENSIONS: cells = columns ; genes = rows
## Grover needs original data format, need to take transpose again
data <- t(data)
dataES <- ExpressionSet(assayData=data) # you must include the 'assayData=' for this to come out correctly

## diffusion map
dm <- DiffusionMap(dataES)

## sigmas
sigmas <- find_sigmas(dataES) 
```

2D Diffusion Map  
```{r diffmap-plot2d}
## plot 2d
plot(x=eigenvectors(dm)[,1], y=eigenvectors(dm)[,2],
     xlab="DC1", ylab="DC2",
     col = colors[age],
     pch=20, 
     main = "Diffusion Map, Grover")
legend("topright", inset=c(0,0),
       legend=levels(age),
       col=colors,
       pch=20,
       ncol=1, cex=0.6, pt.cex=1)
```

3D Diffusion Map  
```{r diffmap-plot3d}
## plot 3d
plot3d(x=eigenvectors(dm)[,1], y=eigenvectors(dm)[,2], z=eigenvectors(dm)[,3],
       xlab="DC1", ylab="DC2", zlab="DC3",
       col = colors[age],
       pch=20, 
       main = "Diffusion Map, Grover")
legend3d("bottomright", inset=c(0.1,0.1),
         legend = levels(age),
         pch = 20,
         col = colors,
         ncol=1, cex=1.5)
rglwidget()
```

Sigmas  
```{r diffmap-sigmas}
plot(sigmas)
```

----

## t-SNE (tsne package)

t-SNE using tsne package (from jdonaldson, linked on van der Maaten's website). t-SNE is not working with Grover data in R. It requires far too much memory, we even tried it with supercomputers but it said it needs terabytes of memory. It seems that the destiny package does NOT use the Barnes-Hut approximation which would dramatically speed up t-SNE.

```{r tsne-grover, eval=FALSE, include=FALSE}
## already read in data and converted to matrix
# can take very, very long 
# t-SNE DIMENSIONS: rows=cells, columns=genes
# Grover data needs transpose to have correct dimensions
data <- t(data)
ydata3d <- tsne(data, k=3, max_iter = 200) # ydata = matrix(rnorm(k * n),n)
```

```{r tsne-plot2d, eval=FALSE, include=FALSE}
# 2d t-SNE plot
plot(x=ydata3d[,1], y=ydata3d[,2], 
     col = colors[groupsf], 
     pch=20,
     main = "t-SNE, Grover") 
legend("topright", inset = c(0,0), 
       legend = levels(groupsf), 
       pch = 20, 
       col = colors, 
       ncol=3, cex=0.6, pt.cex=1) 
```
  
```{r tsne-plot3d, eval=FALSE, include=FALSE}
# 3d t-SNE plot 
plot3d(x=ydata3d[,1], y=ydata3d[,2], z=ydata3d[,3], 
       col=colors[groupsf], 
       pch=20,
       main = "t-SNE, Grover")
legend3d("topright", inset=c(0.1,0.1),
         legend = levels(groupsf),  
         pch = 20, 
         col = colors, 
         ncol=3, cex=1.5)
rglwidget()
```

